src/api/parsekinds.py

```diff
@@ -0,0 +1,158 @@
+#!/usr/bin/env python
+#####################
+## genkinds.py
+## Top contributors (to current version):
+##   Makai Mann
+## This file is part of the CVC4 project.
+## Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+## in the top-level source directory and their institutional affiliations.
+## All rights reserved.  See the file COPYING in the top-level source
+## directory for licensing information.
+##
+
+"""
+This script implements KindsParser which
+parses the header file CVC4/src/api/cvc4cppkind.h
+
+The script is aware of the '#if 0' pattern and will ignore
+kinds declared between '#if 0' and '#endif'. It can also
+handle nested '#if 0' pairs.
+"""
+
+from collections import OrderedDict
+
+#################### Default Filenames ################
+DEFAULT_HEADER = 'cvc4cppkind.h'
+
+##################### Useful Constants ################
+OCB = '{'
+CCB = '}'
+SC = ';'
+EQ = '='
+C = ','
+US = '_'
+NL = '\n'
+
+# Enum Declarations
+ENUM_START = 'enum CVC4_PUBLIC Kind'
+ENUM_END = CCB + SC
+
+# Comments and Macro Tokens
+COMMENT = '//'
+BLOCK_COMMENT_BEGIN = '/*'
+BLOCK_COMMENT_END = '*/'
+MACRO_BLOCK_BEGIN = '#if 0'
+MACRO_BLOCK_END = '#endif'
+
+# Format Kind Names
+# special cases for format_name
+_IS = '_IS'
+# replacements after some preprocessing
+replacements = {
+    'Bitvector': 'BV',
+    'Floatingpoint': 'FP'
+}
+
+
+class KindsParser:
+    tokenmap = {
+        BLOCK_COMMENT_BEGIN: BLOCK_COMMENT_END,
+        MACRO_BLOCK_BEGIN: MACRO_BLOCK_END
+    }
+
+    def __init__(self):
+        self.kinds = OrderedDict()
+        self.endtoken = None
+        self.endtoken_stack = []
+        self.in_kinds = False
+
+    def format_name(self, name):
+        '''
+        In the Python API, each Kind name is reformatted for easier use
```

Comment:
Just flagging this -- I don't know if it makes sense to have the kinds in Java match the C++ kinds exactly, or use this renaming.